website/components/remote-plugin-docs/utils/fetch-github-file.js

```diff
@@ -0,0 +1,91 @@
+const fetch = require('isomorphic-unfetch')
+
+const GITHUB_API_TOKEN = process.env.GITHUB_API_TOKEN
```

Comment:
This utility is what lets us fetch files, such as `nav-data.json` and `content.mdx`, from remote plugin repositories via the GitHub API. I've configured the `GITHUB_API_TOKEN` for both `production` and `preview` environments in the [Vercel env variables](https://vercel.com/hashicorp/packer/settings/environment-variables).

🚨 Note that going forward, authors will need to add a valid `GITHUB_API_TOKEN` (ie, a Personal Access Token) to their `.env.local` file in the root of the repository in order for the site to build as expected.

Reply:
> Note that going forward, authors will need to add a valid GITHUB_API_TOKEN (ie, a Personal Access Token) to their .env.local file in the root of the repository in order for the site to build as expected.

Is this warning exclusively for folks wanting to run the website locally? If so I think that is okay and we need to make sure it is captured in the website README. 

What happens if we can't fetch the remote location. Do we get a blank page or error?

Reply:
@nywilken yep, this is exclusively for folks wanting to run the website locally. PR deploy previews, and of course the production website, will have access to a `GITHUB_API_TOKEN` from our Vercel env. I'll add a note to the README to this effect, that's a great call! 🙇 

Locally, what happens is the build will throw an error. In local development, the error will look something like:

![CleanShot 2021-02-26 at 16 35 54](https://user-images.githubusercontent.com/4624598/109357800-d15a3980-7850-11eb-8ed1-da47efafef62.png)

Reply:
Cool - very description error messaging that tells the user what to do next. Love it!

Reply:
Do we get an error trying to view any page without a local token, or just the pages that we need the token for? Ideally it would be the former 😀

Reply:
The latter, unfortunately - in this specific case, all `docs` pages, since they all rely on the fetch requests to know which URLs need to be rendered in `getStaticPaths`.

As discussed in Slack, I'll look into a way we can avoid this - I think we _want_ these errors to throw in production, to alert us to missing content. So I'll target dev with NODE_ENV, and if GITHUB_API_TOKEN isn't defined, I'll see if I can add a big red box to the top of all `docs` pages if it's NOT defined. (If it is defined, but is invalid, or if GitHub is down, then in local dev folks would still see the error above).

Reply:
Adding here that Github's public REST api allows to do a few requests for free without a token: maybe it could be nice in dev mode, because a dev will go and check their files.

```shellsession
> curl https://api.github.com/repos/azr/packer-plugin-amazon/contents/main.go | jq -r .content | base64 --decode
package main

import (
	"fmt"
	"os"

	"github.com/hashicorp/packer/builder/amazon/ebs"
	"github.com/hashicorp/packer/builder/amazon/ebssurrogate"
	"github.com/hashicorp/packer/builder/amazon/ebsvolume"
	"github.com/hashicorp/packer/builder/osc/chroot"
	"github.com/hashicorp/packer/packer-plugin-sdk/plugin"
	amazonimport "github.com/hashicorp/packer/post-processor/amazon-import"
)

func main() {
	pps := plugin.NewSet()
	pps.RegisterBuilder("ebs", new(ebs.Builder))
	pps.RegisterBuilder("chroot", new(chroot.Builder))
	pps.RegisterBuilder("ebssurrogate", new(ebssurrogate.Builder))
	pps.RegisterBuilder("ebsvolume", new(ebsvolume.Builder))
	pps.RegisterPostProcessor("import", new(amazonimport.PostProcessor))
	err := pps.Run()
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
}
```

Reply:
@azr I love that idea - being able to show that remote previews without a `GITHUB_API_TOKEN` would be huge. 

We also have some ideas about how this could work outside the `packer` repo - eg one is that we could allow running `npx -p @hashicorp/packer-docs-artifacts preview` within plugin repositories (or really any directory with a `docs` folder), and seeing a live, accurate preview of just that individual plugin's docs and sidebar, with live reload. This would remove the need for API requests altogether (at least just for the isolated preview use case).

For now I've taken a slightly less ideal route.... but hopefully still better than just throwing an error and requiring a `GITHUB_API_TOKEN` to preview _anything_ on the `docs` route. If there's no `GITHUB_API_TOKEN`, **and** we detect things are running locally (via our `VERCEL_ENV` var), then we'll show a nice clear message on `docs` routes rather than throwing an error:

<img width="1536" alt="CleanShot 2021-03-02 at 19 00 53@2x" src="https://user-images.githubusercontent.com/4624598/109732280-e6093b00-7b8a-11eb-9f5b-6f3d24afbfcc.png">

This way, local builds won't immediately break and force the addition of a `GITHUB_API_TOKEN` when we ship this feature ☺️  CC @jescalan @nywilken as this is an update to the earlier comment

> Note: in production, and in deploy previews, we will still throw an error if we can't access remote content.